java/com/google/flatbuffers/Table.java

```diff
@@ -71,20 +83,34 @@ protected int __indirect(int offset) {
    * @return Returns a `String` from the data stored inside the FlatBuffer at `offset`.
    */
   protected String __string(int offset) {
+    CharsetDecoder decoder = UTF8_DECODER.get();
+    decoder.reset();
+
     offset += bb.getInt(offset);
-    if (bb.hasArray()) {
-      return new String(bb.array(), bb.arrayOffset() + offset + SIZEOF_INT, bb.getInt(offset),
-                        FlatBufferBuilder.utf8charset);
-    } else {
-      // We can't access .array(), since the ByteBuffer is read-only,
-      // off-heap or a memory map
-      ByteBuffer bb = this.bb.duplicate().order(ByteOrder.LITTLE_ENDIAN);
-      // We're forced to make an extra copy:
-      byte[] copy = new byte[bb.getInt(offset)];
-      bb.position(offset + SIZEOF_INT);
-      bb.get(copy);
-      return new String(copy, 0, copy.length, FlatBufferBuilder.utf8charset);
+    ByteBuffer src = bb.duplicate().order(ByteOrder.LITTLE_ENDIAN);
```

Comment:
No, it's because if you don't duplicate it, it's not thread safe to multiple concurrent readers.  So in-between changing its position/limit and then resetting, any other reader that is using the same buffer would be susceptible to getting strange exceptions such as buffer underflow/overflow exceptions.